### PR TITLE
fix: handle reserved var keyword and url definition

### LINKED ISSUE #12: `var` keyword will generate invalid output
Upcoming Defold v1.9.3 uses the `var` keyword as a parameter in its documentation: https://defold.com/ref/beta/types/#types.is_hash:var

Use of this reserved keyword is invalid in TypeScript: `SyntaxError: 'var' is not allowed as a parameter name.`

We should probably detect the `var` parameter and change it to `v`.

### LINKED ISSUE #13: `url` definition is missing properties
Currently we generate `url` as `{}`

The full shape of `url` is this:
```
declare type url = {
	socket: hash;
	path: hash;
	fragment: hash | undefined;
};
```

### PR DESCRIPTION
**Pull request by thinknathan**
Fixes #12 
Sort of fixes #13 although the actual definition for url's `fragment` should be `hash | undefined`. I wasn't sure how to create the union with `undefined`, so the output is just `hash`.

### CODE REVIEW

File: src/gen/parser.ts [hunk 1]
```diff
@@ -66,7 +66,10 @@ function parseName(
   const maybeOptional = /\[(.+)\]/.exec(input);
   const optional = maybeOptional && maybeOptional?.length > 1 ? true : false;
   const rawName = optional && maybeOptional ? maybeOptional[1] : input;
-  const name = rawName.replace(/^[^a-zA-Z_$]|[^0-9a-zA-Z_$]/g, '_');
+  const name = rawName
+    .replace(/^[^a-zA-Z_$]|[^0-9a-zA-Z_$]/g, '_')
+    // Replace reserved keyword `var`
+    .replace('var', 'v');
```
Comment (thejustinwalsh) [inline src/gen/parser.ts:72]: Seems like a reasonable solution. The variable name doesn't appear to be important to the context of its function in the calls that use var as a parameter name.

File: src/types/builtIns.ts [hunk 2]
```diff
@@ -14,7 +14,11 @@ const types: Array<IDocTypes> = [
         name: 'url',
         unions: [],
         intersections: [],
-        definition: {},
+        definition: {
+          socket: EDocParamType.Hash,
+          path: EDocParamType.Hash,
+          fragment: EDocParamType.Hash,
```
Comment (thejustinwalsh) [inline src/types/builtIns.ts:20]: Should this be `hash | unknown`? I don't recall if the generator has a way to represent unknown vs any at the moment.

**Review by thejustinwalsh — APPROVED** [review]
Unsure about what the object structure of a fragment would be in the case that it is not a hash. So looks good from my perspective.

**Comment by thinknathan** [comment]
@thejustinwalsh Thanks for the review!

File: src/types/builtIns.ts (hunk 2 above)
Comment (thinknathan) [reply src/types/builtIns.ts:20]: Yeah, it's probably fine to leave as just `hash`. The `any` keyword would override the union.

**Comment by thejustinwalsh** [comment]
@thinknathan all we should have to do is bump the version and commit a tag that matches to publish this via the workflow.

i.e. https://github.com/ts-defold/type-gen/commit/0d5e88bcdc7ea3e6f0981f6f98247e341d80fbf6
```sh
git tag v0.5.4
git push origin main
git push origin v0.5.4
```

**Pull request merged**